man/man1/fzctl.1
.Dd 2024-05-01
.Dt fzctl 1
.Os
.Sh NAME
.Nm fzctl
.Nd Run commands against an fz daemon.
.Sh SYNOPSIS
.Nm
.Op options
.Ar list
.Op ok|fail|unknown
.Nm
.Op options
.Ar show
.Ar task
.Op task...
.Sh DESCRIPTION
Control an
.Nm fz
daemon, or query its internal state.
.Pp
.Bl -tag -width Ds
.It Fl c, Li --config
The
.Nm fz
configuration file. It will connect to the
.Xr listen_address
unless the --host flag was supplied.
.It Fl H, Li --host
The IP address and port of the
.Nm fz
daemon to connect to. The default value is 127.0.0.1:5891.
.It Fl V, Li --version
Print the version information.
.El
.Sh ENVIRONMENT
Some settings can also be supplied with environment variables, but if a command arguments is provided it will take precedence.
.Bl -tag -width Ds
.It Ic FZ_LISTEN
The network address that the
.Xr fz 1
daemon is listening on.
.El
.Sh SEE ALSO
.Xr fz 1 ,
.Xr flamingzombies.toml 5


CHANGELOG.md
# CHANGELOG

## v0.0.21

- Add `OPEN_GATES` to the environment of notifiers.
- Add `task/tls_expiration` plugin.
- Add `task/loadavg` plugin.
- Tag the messages from `ntfy`.

## v0.0.20

- Many improvements to the docs and the interface.
- Removed dependency on logrus so that it can cross-compile for OpenBSD 7.5.
- Support gatesets to combine ANDing or ORing the results of gates.
- Improvements to the configtest
- Add `task/fz_errors` plugin
- Add `LAST_OK`, `LAST_FAIL` and `LAST_NOTIFICATION` to the environment of the gates.
- Support delayed notifications.
- Support re-raising notifications.
- Support detecting when a task is flapping.

## v0.0.19

- Add fzctl to check the daemon state
- Add scripts for testing HTTP
- Slight improvements to the man pages

## v0.0.18

- Release a tarball of plugins

## v0.0.17

- Test a release

## v0.0.16

- Simplify the release logic

## v0.0.15

- Improvements to the generated OpenBSD artifact
- Add a configtest mode.

## v0.0.14

- Remove dependency on goreleaser

## v0.0.13

- Fix crashes on OpenBSD

## v0.0.10

- License it under BSD-2-Clause
- Start a man page

## v0.0.7

- Add option parsing

## v0.0.5

- Another test release


cmd/fzctl/fzctl.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net"
	"os"
	"time"

	"nullprogram.com/x/optparse"
)

const VERSION = "v0.0.20"

type taskData struct {
	Name           string    `json:"name"`
	State          string    `json:"state"`
	LastRun        time.Time `json:"last_run"`
	LastOk         time.Time `json:"last_ok"`
	Measurements   []bool    `json:"measurements"`
	ExecutionCount int       `json:"execution_count"`
	OKCount        int       `json:"ok_count"`
	FailCount      int       `json:"fail_count"`
	ErrorCount     int       `json:"error_count"`
}

var tasks []taskData
var cmd = "list"
var subcmds = []string{}

func init() {
	if os.Getenv("FZ_LISTEN") == "" {
		os.Setenv("FZ_LISTEN", "127.0.0.1:5891")
	}

	options := []optparse.Option{
		{"help", 'h', optparse.KindNone},
		{"host", 'H', optparse.KindRequired},
		{"version", 'V', optparse.KindNone},
	}

	results, _, err := optparse.Parse(options, os.Args)
	if err != nil {
		log.Fatal(err)
	}

	optCount := 0
	for _, result := range results {
		switch result.Long {
		case "help":
			usage()
			return
		case "host":
			os.Setenv("FZ_LISTEN", result.Optarg)
			optCount += 2
		case "version":
			fmt.Printf("fzctl %s\n", VERSION)
			os.Exit(0)
		}
	}

	if len(os.Args) > optCount+1 {
		cmd = os.Args[optCount+1]
	}

	if len(os.Args) > optCount+2 {
		subcmds = os.Args[optCount+2:]
	}
}

func main() {
	switch cmd {
	case "list":
		list()
	case "show":
		show()
	case "help":
		usage()
	default:
		fmt.Fprintf(os.Stderr, "Unknown command: %s\n", cmd)
		os.Exit(1)
	}
}

func list() {
	for _, s := range subcmds {
		if s != "ok" && s != "fail" && s != "unknown" {
			fmt.Fprintf(os.Stderr, "Unsupported subcommand: %s\n", s)
			fmt.Fprintln(os.Stderr, "  allowed values are 'ok', 'fail', or 'unknown'.")
			os.Exit(1)
		}
	}

	fetchTasks()

	if len(subcmds) == 0 {
		subcmds = []string{"ok", "fail", "unknown"}
	}
	for _, s := range subcmds {
		for _, t := range tasks {
			if t.State == s {
				fmt.Printf("%-40s\t%s\n", t.Name, t.State)
			}
		}
	}
}

func show() {
	if len(subcmds) == 0 {
		fmt.Fprintln(os.Stderr, "No task was provided.")
		os.Exit(1)
	}

	fetchTasks()

	for i, s := range subcmds {
		for _, t := range tasks {
			if t.Name == s {
				if i != 0 {
					fmt.Println("")
				}

				fmt.Printf("%-20s%s\n", "name:", t.Name)
				fmt.Printf("%-20s%s\n", "state:", t.State)
				fmt.Printf("%-20s%s ago\n", "last execution:", time.Now().Sub(t.LastRun))
				fmt.Printf("%-20s%s ago\n", "last success:", time.Now().Sub(t.LastOk))
				fmt.Printf("%-20s%d\n", "executions:", t.ExecutionCount)
				fmt.Printf("%-20s%d\n", "failures:", t.FailCount)
				fmt.Printf("%-20s%d\n", "errors:", t.ErrorCount)
			}
		}
	}
}

func fetchTasks() {
	data, err := readFromNetwork(fmt.Sprintf("%s", os.Getenv("FZ_LISTEN")))
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %s\n", err)
		os.Exit(1)
	}

	scanner := bufio.NewScanner(bytes.NewReader(data))
	for scanner.Scan() {
		var t taskData
		err = json.Unmarshal(scanner.Bytes(), &t)
		if err != nil {
			fmt.Fprintf(os.Stderr, "error: %s\n", err)
			os.Exit(1)
		}

		tasks = append(tasks, t)
	}
	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %s\n", err)
		os.Exit(1)
	}
}

func readFromNetwork(addr string) ([]byte, error) {
	conn, err := net.Dial("tcp", addr)
	if err != nil {
		return []byte{}, err
	}
	defer conn.Close()

	buf := make([]byte, 1)
	var data []byte

	for {
		_, err := conn.Read(buf)
		if err != nil {
			if err == io.EOF {
				break
			}
			return []byte{}, err
		}

		data = append(data[:], buf[:]...)
	}

	return data, nil

	//fmt.Println("Read", len(data), "bytes from the connection:", string(data[:len(data)]))
}

func usage() {
	fmt.Println("Usage:")
	fmt.Println("  fzctl [OPTIONS] <command> [subcommand]")
	fmt.Println("")
	fmt.Println("Options:")
	fmt.Println("  -H, --host <ip:port>        Host to connect to")
	fmt.Println("  -V, --version               Version")
	fmt.Println("")
	fmt.Println("Commands:")
	fmt.Println("  help                        This help")
	fmt.Println("  list [ok] [fail] [unknown]  List the running tasks")
	fmt.Println("  show <task>                 Show the details of a task")
	os.Exit(0)
}
